File: typescript/problem_31.ts
export function editDistance(a: string, b: string): number {
    const m = a.length, n = b.length;
    const dp: number[][] = Array.from({ length: m + 1 }, (_, i) =>
        Array.from({ length: n + 1 }, (_, j) => (i === 0 ? j : j === 0 ? i : 0))
    );
    for (let i = 1; i <= m; i++)
        for (let j = 1; j <= n; j++)
            dp[i][j] = a[i-1] === b[j-1]
                ? dp[i-1][j-1]
                : 1 + Math.min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1]);
    return dp[m][n];
}

export function problem31(): void {
    console.log("Problem 31:");
    if (editDistance("kitten", "sitting") !== 3) throw new Error("assertion failed");
    if (editDistance("", "abc")           !== 3) throw new Error("assertion failed");
    if (editDistance("abc", "")           !== 3) throw new Error("assertion failed");
    if (editDistance("abc", "abc")        !== 0) throw new Error("assertion failed");
    if (editDistance("a", "b")            !== 1) throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: lua/problem_31.lua
local function editDistance(a, b)
    local m, n = #a, #b
    local dp = {}
    for i = 0, m do
        dp[i] = {}
        for j = 0, n do
            if i == 0 then dp[i][j] = j
            elseif j == 0 then dp[i][j] = i
            else dp[i][j] = 0 end
        end
    end
    for i = 1, m do
        for j = 1, n do
            if a:sub(i,i) == b:sub(j,j) then
                dp[i][j] = dp[i-1][j-1]
            else
                dp[i][j] = 1 + math.min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1])
            end
        end
    end
    return dp[m][n]
end

local function problem31()
    print("Problem 31:")
    assert(editDistance("kitten", "sitting") == 3, "assertion failed")
    assert(editDistance("", "abc")           == 3, "assertion failed")
    assert(editDistance("abc", "")           == 3, "assertion failed")
    assert(editDistance("abc", "abc")        == 0, "assertion failed")
    assert(editDistance("a", "b")            == 1, "assertion failed")
    print("All Tests Passed")
end

return { problem31 = problem31 }
